build-system/babel-plugins/babel-plugin-transform-log-methods/index.js

```diff
@@ -0,0 +1,221 @@
+      let i = 0;
+      for (; i < node.quasis.length - 1; i++) {
+        quasied += node.quasis[i].value.cooked;
+        quasied += buildMessageSprintf(node.expressions[i], interpolationArgs);
+      }
+      quasied += node.quasis[i].value.cooked;
+      return quasied;
+    }
+
+    interpolationArgs.push(node);
+    return '%s';
+  }
+
+  /**
+   * @param {!Node} node
+   * @return {../../log-module-metadata.LogMethodMetadataDef}
+   */
+  function getTransformableCalleeMeta({callee}) {
+    if (assertAliases.some(name => t.isIdentifier(callee, {name}))) {
+      return transformableMethods.assert;
+    }
+    // is method().call().
+    if (!t.isMemberExpression(callee) || !t.isCallExpression(callee.object)) {
+      return;
+    }
+    // is either dev() or user() object.
+    const singletonCallee = callee.object.callee;
+    if (
+      !singletonFunctions.some(name => t.isIdentifier(singletonCallee, {name}))
+    ) {
+      return;
+    }
+    if (!t.isIdentifier(callee.property)) {
+      return;
+    }
+    // callee property is transformable method, otherwise undefined.
+    return transformableMethods[callee.property.name];
+  }
+
+  return {
+    /** Resolves plugin config. */
+    pre() {
+      // Temporary option to not replace call arguments, but still output the
+      // table to keep build code and infra independent from rollout.
+      const {replaceCallArguments = true} = this.opts;
+      shouldReplaceCallArguments = replaceCallArguments;
+
+      // Configurable to isolate test output.
+      messagesPath = relativeToRoot(this.opts.messagesPath || extractedPath);
+
+      // Read table.
+      messages = fs.readJsonSync(messagesPath, {throws: false}) || {};
+      nextMessageId = Object.keys(messages).length;
+    },
+    visitor: {
+      /** Write table on every exit. */
+      Program: {
+        exit() {
+          fs.outputJsonSync(messagesPath, messages, {spaces: 2});
+        },
+      },
+      /**
+       * Visits call expressions for known log assertion/error methods.
+       * @param {*} path babel.path
+       */
+      CallExpression({node}) {
+        const meta = getTransformableCalleeMeta(node);
+        if (!meta) {
+          return;
+        }
+
+        const {extractMessages, messageArgPos} = meta;
+        if (!extractMessages) {
+          return;
+        }
+
+        const messageArg = node.arguments[messageArgPos];
+        if (!messageArg) {
+          return;
+        }
+
+        // Recursively construct sprintf template from arguments.
+        const templateArgs = [];
+        const message = buildMessageSprintf(messageArg, templateArgs);
+
+        // Bounce when indirection does nothing (±1 byte delta). (Also catches
+        // the case when its template is exactly `%s`).
+        if (message.length <= roughNestedLength) {
+          return;
+        }
+
+        const idLiteral = t.stringLiteral(getOrCreateMessageId(message));
+
+        if (!shouldReplaceCallArguments) {
```

Comment:
since this is assigned in the `pre` can we move this higher? or are there processes that we want to execute and that's why we evaluate this late?

Reply:
Yeah, we want to output the extracted table but not actually transform the code. We'll run the transform once everything on the server side is ready.